samples/EventImporter/Provider.Tests/EventImporterProviderTests.cs

```diff
@@ -0,0 +1,145 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Moq;
+using PactNet;
+using PactNet.AspNetCore.Messaging.Options;
+using PactNet.Infrastructure.Outputters;
+using PactNet.Native;
+using Provider.Api;
+using Provider.Api.Controllers;
+using Provider.Domain;
+using Provider.Domain.Models;
+using Provider.Infrastructure;
+using Xunit;
+using Xunit.Abstractions;
+using EventHandler = Provider.Domain.Handlers.EventHandler;
+
+namespace Provider.Tests
+{
+    public class EventImporterProviderTests : IClassFixture<EventImporterFixture>
+    {
+        private readonly EventImporterFixture fixture;
+        private readonly ITestOutputHelper output;
+        private readonly MessagingVerifierOptions options;
+        private readonly Scenarios scenarios;
+
+        public EventImporterProviderTests(EventImporterFixture fixture, ITestOutputHelper output)
+        {
+            this.fixture = fixture;
+            this.output = output;
+            this.options = fixture.GetOptions();
+            this.scenarios = fixture.Scenarios;
+        }
+
+        [Fact]
+        public void EnsureEventApiHonoursPactWithConsumer()
+        {
+            var config = new PactVerifierConfig
+            {
+                Outputters = new List<IOutput>
+                {
+                    new XUnitOutput(output)
+                }
+            };
+
+            //C
+            string pactPath = Path.Combine("..",
+                                           "..",
+                                           "..",
+                                           "..",
+                                           "Consumer.Tests",
+                                           "pacts",
+                                           "Event API Consumer V3 Message-Event API V3 Message.json");
+
+            //Setting up scenarios for messaging simulation.
+            //If other provider tests, the unique constraint would need to be handled (the ClearScenarios method of the Scenarios class is available)
+            SetupEventImporterScenarios();
```

Comment:
I'm generally confused about this whole thing works, perhaps because it's backwards from the usual request/response type. So the consumer has already defined the pact file, and now the provider is effectively setting up a sort of provider state using the magic strings inside that pact file?

That seems quite brittle, because new interactions created by the consumer immediately need code changes in the provider otherwise it's going to throw exceptions and fail the test.

Perhaps I'm just not understanding how this works, but I've read through it twice now and I don't really get it.

Reply:
You've got to think of the whole messaging approach.
The workflow is indeed reversed in the real-life API than it would be for HTTP interaction model.

1) The API producing the resource first pushes it to the queue.
2) The API that subscribed to the queue is reading the queue afterwards asynchronously. 

The problem is in PACT, you cannot reverse the workflow since it's consumer driven.

Therefore, you need a way to "emulate" the message pushed to validate it against the consumer contract. Very likely, the API reading the queue won't have any public endpoint to hit like an REST API would, so we need a way to set up messages that would otherwise be pushed to the queue.

If you read my commented code in the event importer test class, you can see the two approaches to validate messages. One by only checking a typed object and the other one to actually execute code and by a way of mocking, get the message before it's sent to the queue and compare it with the PACT contract.

Reply:
The `AddScenario` builder actually reads quite well, it's the `WithContent` block that I'm finding hardest to follow (but probably because it's .NET specifics), but I think it does what the others do. Basically you need to get the piece of code that normally emits an event to a queue (or queue-like thing), and return it instead. 

This usually means you need to separate the code that generates the message from the bit that publishes to the queue. We care about the former. But in this case, the separation isn't there and you're intercepting the behaviour. This is probably going to be useful for those people that have good separation of concerns. It would probably be ideal to have a "here's the best practice" version of it though (doesn't need to be this PR).

For reference, this is what a message pact test looks like in 

* Java: https://docs.pact.io/implementation_guides/jvm/provider/junit/#example-of-message-test
* Ruby: https://github.com/pact-foundation/pact-message-ruby#provider
* JS: https://github.com/pact-foundation/pact-js/#consumer

@adamrodger see also this [doc](https://docs.pact.io/getting_started/how_pact_works#non-http-testing-message-pact) on "how" we test messages. It might be largely unsatisfying, but it's actually quite powerful.

I don't know the state of .NET and frameworks that wrap message queue type things, but because there is no single standard (like HTTP) we haven't taken the path of "support all of the types". This also means we can support use cases like dropping a JSON file to an S3 bucket - the expense is a fair amount of abstraction.

Reply:
Very nice comment @mefellows. 
If you look at the class EventImporterProviderTests and the methode SetupEventImporterScenarios in the sample app EventImporter I actually wrote the two approaches. Granted I didn't separate the code inside the WithContent like adding it to another class or method.

The problem I have with your example in java for instance is that you are only checking the type of the message and not the piece of code that publishes the message intercepting the message before sending it to the queue.

The WithContent method has two overloads. One with the message and one with a handle method. Would it make more sense to have the method named like this ?
- PublishMessage(dynamic message)
- PublishMessageWith(Func<dynamic> handleMessage)